### PR TITLE
Update Bonsai docs

### PR DESCRIPTION
**Pull request by capossele**
- Update Bonsai docs
- Update Bonsai Ethereum Relay README

### CODE REVIEW

File: bonsai/ethereum-relay/README.md [hunk 1]
```diff
@@ -9,77 +9,118 @@ The picture below shows a simplified overview of how users can integrate Bonsai
...
-6. Once Bonsai has generated a proof of execution, the Bonsai Relay will forward this proof along with the result of the computation to the Bonsai Relay Smart Contract.
-7. This triggers a verification of the proof on-chain, and only upon successful verification, the result of the computation will be forwarded to the original requester Smart Contract `A` via the `invoke_callback` function.
-
-### REST API
-
-As an alternative to sending a `Callback request` from Ethereum as described by step 5, the request can be sent directly to the Bonsai Relay via an HTTP REST API.
-Then, the remaining steps will flow as above. The following example explains how to do that.
+1. Deploy the `Bonsai Relay Contract` to Ethereum using address `0xB..`.
+2. Launch the `Bonsai Relayer`, setting `--contract-address` to `0xB..`.
+3. To delegate off-chain computation for Smart Contract `A` to Bonsai, register its `Image` or `ELF` (the binary executing the computation on RISC Zero ZKVM) with Bonsai.
```
Comment (tzerrell) [inline bonsai/ethereum-relay/README.md:28]: I was a bit confused about whether "Smart Contract `A`" was a contract mentioned before that I was supposed to know something about. Maybe this is clearer?
```suggestion
3. To delegate off-chain computation for a new Smart Contract `A` to Bonsai, register its `Image` or `ELF` (the binary executing the computation on RISC Zero ZKVM) with Bonsai.
```

File: xtask/Cargo.toml [hunk 2]
```diff
@@ -13,7 +13,7 @@ risc0-fault = { path = "../risc0/fault" }
 risc0-zkp = { workspace = true }
 risc0-zkvm = { workspace = true, features = ["prove"] }
 risc0-zkvm-methods = { path = "../risc0/zkvm/methods" }
-tempfile = "3.5"
+tempfile = "3.3"
```
Comment (tzerrell) [inline xtask/Cargo.toml:16]: I'm surprised by this version downgrade -- if there's a reason I'm not concerned, but I'm calling attention to it in case it was an accident.

File: website/api_versioned_docs/version-0.19/bonsai/blockchain-zkvm-guide.md
Comment (tzerrell) [inline website/api_versioned_docs/version-0.19/bonsai/blockchain-zkvm-guide.md:1]: It looks like you duplicated the main content of this file, rather than overwriting it -- there are now 2 copies of most paragraphs.

File: website/static/img/bonsai_ethereum.png
Comment (tzerrell) [inline website/static/img/bonsai_ethereum.png:1]: I like this new diagram! (and the others as well)

File: website/api_versioned_docs/version-0.19/bonsai/bonsai-on-eth.md [hunk 3]
```diff
@@ -2,18 +2,133 @@
...
+
+The Bonsai Ethereum Relay provides both an _off-chain_ (vie the HTTP REST API of the `Bonsai Relayer`) and an _on-chain_ (via the [Bonsai Relay Contract]) interface to send `Callback requests`.
+
+### Off-chain
+
+A typical flow works as follows:
+
+1. Deploy the `Bonsai Relay Contract` to Ethereum using address `0xB..`.
+2. Launch the `Bonsai Relayer`, setting `--contract-address` to `0xB..`.
+3. To delegate off-chain computation for Smart Contract `A` to Bonsai, register its `Image` or `ELF` (the binary executing the computation on RISC Zero ZKVM) with Bonsai.
```
Comment (tzerrell) [inline website/api_versioned_docs/version-0.19/bonsai/bonsai-on-eth.md:32]: This is the same as the README, and I have the same suggestion that might improve clarity
```suggestion
3. To delegate off-chain computation for a new Smart Contract `A` to Bonsai, register its `Image` or `ELF` (the binary executing the computation on RISC Zero ZKVM) with Bonsai.
```

File: website/api_versioned_docs/version-0.19/bonsai/bonsai-on-eth.md [hunk 4]
```diff
@@ -2,18 +2,133 @@
...
+5. Initiate a `Callback request` through the `Bonsai Relayer`'s off-chain REST API. This request is sent to Bonsai.
+6. After Bonsai produces an execution proof, the `Bonsai Relayer` relays the proof and computation result to the `Bonsai Relay Contract`.
+7. Ethereum verifies the proof on-chain. If validated, Smart Contract `A` receives the computation result via the `invoke_callback` function.
+
+#### Example
+
+The following example assumes that the `Bonsai Relayer` is up and running with the server API enabled,
+and that the memory image of your `ELF` is already registered against Bonsai with a given `IMAGE_ID` as its identifier.
+
+```rust,no_run
```
Comment (tzerrell) [inline website/api_versioned_docs/version-0.19/bonsai/bonsai-on-eth.md:43]: It is awkward from a maintenance standpoint to have this much code and writing duplicated between the README and the website. If you're onboard from doing maintenance I guess it's ok, but de-duplicating docs would be nice if it feels reasonable to have one reference the other.

File: website/api_versioned_docs/version-0.19/bonsai/bonsai-overview.md [hunk 5]
```diff
@@ -7,42 +7,43 @@ slug: ./
 Bonsai enables boundless computation on any blockchain.
 With Bonsai, instead of generating proofs on your own hardware, you can upload the program you need proven and request that Bonsai generates proofs for you as needed.
 
-Bonsai is highly parallelized and highly performant. To get a taste of what you can do with Bonsai, check out our [Governance Showcase], the [Bonfire Wallet] demo from the Hash Cloak team, and our twitter thread about using [Bonsai as a zk coprocessor].
+Bonsai is highly parallelized, highly performant and can horizontally scale. To get a taste of what you can do with Bonsai, check out [Zeth], our[Governance Showcase], the [Bonfire Wallet] demo from the Hash Cloak team, and our twitter thread about using [Bonsai as a zk coprocessor].
```
Comment (tzerrell) [inline website/api_versioned_docs/version-0.19/bonsai/bonsai-overview.md:10]: Typo
```suggestion
Bonsai is highly parallelized, highly performant and can horizontally scale. To get a taste of what you can do with Bonsai, check out [Zeth], our [Governance Showcase], the [Bonfire Wallet] demo from the Hash Cloak team, and our twitter thread about using [Bonsai as a zk coprocessor].
```

File: website/api/bonsai/bonsai-overview.md [hunk 6]
```diff
@@ -7,42 +7,43 @@ slug: ./
 Bonsai enables boundless computation on any blockchain.
 With Bonsai, instead of generating proofs on your own hardware, you can upload the program you need proven and request that Bonsai generates proofs for you as needed.
 
-Bonsai is highly parallelized and highly performant. To get a taste of what you can do with Bonsai, check out our [Governance Showcase], the [Bonfire Wallet] demo from the Hash Cloak team, and our twitter thread about using [Bonsai as a zk coprocessor].
+Bonsai is highly parallelized, highly performant and can horizontally scale. To get a taste of what you can do with Bonsai, check out [Zeth], our[Governance Showcase], the [Bonfire Wallet] demo from the Hash Cloak team, and our twitter thread about using [Bonsai as a zk coprocessor].
```
Comment (tzerrell) [inline website/api/bonsai/bonsai-overview.md:10]: Typo
```suggestion
Bonsai is highly parallelized, highly performant and can horizontally scale. To get a taste of what you can do with Bonsai, check out [Zeth], our [Governance Showcase], the [Bonfire Wallet] demo from the Hash Cloak team, and our twitter thread about using [Bonsai as a zk coprocessor].
```

File: website/api/bonsai/bonsai-on-eth.md [hunk 7]
```diff
@@ -2,18 +2,133 @@
...
+
+The Bonsai Ethereum Relay provides both an _off-chain_ (vie the HTTP REST API of the `Bonsai Relayer`) and an _on-chain_ (via the [Bonsai Relay Contract]) interface to send `Callback requests`.
+
+### Off-chain
+
+A typical flow works as follows:
+
+1. Deploy the `Bonsai Relay Contract` to Ethereum using address `0xB..`.
+2. Launch the `Bonsai Relayer`, setting `--contract-address` to `0xB..`.
+3. To delegate off-chain computation for Smart Contract `A` to Bonsai, register its `Image` or `ELF` (the binary executing the computation on RISC Zero ZKVM) with Bonsai.
```
Comment (tzerrell) [inline website/api/bonsai/bonsai-on-eth.md:32]: This is the same as the README, and I have the same suggestion for language that might be clarifying.
```suggestion
3. To delegate off-chain computation for the new Smart Contract `A` to Bonsai, register its `Image` or `ELF` (the binary executing the computation on RISC Zero ZKVM) with Bonsai.
```

File: website/api/bonsai/bonsai-on-eth.md [hunk 8]
```diff
@@ -2,18 +2,133 @@
...
-1. Deployed a [smart contract] on chain that calls the relay contract
-2. Uploaded a compiled [program] for the zkVM to the Bonsai proving service
+1. Write the [program] you want proven and upload it to Bonsai
+2. Write the on-chain part of your application and deploy it
+
+## Interfaces
+
+The Bonsai Ethereum Relay provides both an _off-chain_ (vie the HTTP REST API of the `Bonsai Relayer`) and an _on-chain_ (via the [Bonsai Relay Contract]) interface to send `Callback requests`.
+
+### Off-chain
```
Comment (tzerrell) [inline website/api/bonsai/bonsai-on-eth.md:26]: The amount of duplicated code and writing between here an the readme worries me a little. If it's possible to de-duplicate this by having one reference the other it would probably reduce maintenance needs.

File: .github/workflows/main.yml [hunk 9]
```diff
@@ -179,6 +179,7 @@ jobs:
           # Recursive checkout required to build Solidity contracts and bindings.
           submodules: recursive
       - uses: ./.github/actions/rustup
+      - uses: risc0/foundry-toolchain@2fe7e70b520f62368a0e3c464f997df07ede420f
```
Comment (flaub) [inline .github/workflows/main.yml:182]: Why do we need this?

File: website/api/bonsai/bonsai-on-eth.md [hunk 10]
```diff
@@ -25,3 +140,5 @@ You may also want to check out our [Bonsai Quick Start](quickstart.md) page.
 [Bonsai Foundry Template]: https://github.com/risc0/bonsai-foundry-template
 [smart contract]: https://github.com/risc0/bonsai-foundry-template/tree/main/contracts
 [program]: https://github.com/risc0/bonsai-foundry-template/tree/main/methods/guest/src/bin
+[Bonsai Relay Contract]: https://github.com/risc0/risc0/bonsai/ethereum/contracts/BonsaiRelay.sol
```
Comment (flaub) [inline website/api/bonsai/bonsai-on-eth.md:143]: This link is broken

File: website/api/bonsai/bonsai-on-eth.md [hunk 11]
```diff
@@ -2,18 +2,133 @@
...
+5. Initiate a `Callback request` through the `Bonsai Relayer`'s off-chain REST API. This request is sent to Bonsai.
+6. After Bonsai produces an execution proof, the `Bonsai Relayer` relays the proof and computation result to the `Bonsai Relay Contract`.
+7. Ethereum verifies the proof on-chain. If validated, Smart Contract `A` receives the computation result via the `invoke_callback` function.
+
+#### Example
+
+The following example assumes that the `Bonsai Relayer` is up and running with the server API enabled,
+and that the memory image of your `ELF` is already registered against Bonsai with a given `IMAGE_ID` as its identifier.
+
+```rust,no_run
```
Comment (flaub) [inline website/api/bonsai/bonsai-on-eth.md:43]: Use a space:
```suggestion
```rust no_run
```

File: website/api_versioned_docs/version-0.19/bonsai/bonsai-on-eth.md [hunk 12]
```diff
@@ -2,18 +2,133 @@
...
```
Comment (flaub) [inline website/api_versioned_docs/version-0.19/bonsai/bonsai-on-eth.md:43]:
```suggestion
```rust no_run
```

File: .github/workflows/main.yml (hunk 9 above)
Comment (capossele) [reply .github/workflows/main.yml:182]: I've removed the `ignore` in the rust example of the Ethereum Relay's README. That imports the Ethereum Relay that in turns imports the ethereum contracts, that requires `forge build` when building the contracts.
As Tim pointed out though, since we now have the same example in the docs, I can remove it from there and instead refer to the documentation for that particular example

File: website/api/bonsai/bonsai-on-eth.md (hunk 10 above)
Comment (capossele) [reply website/api/bonsai/bonsai-on-eth.md]: Good catch!

File: website/api/bonsai/bonsai-on-eth.md (hunk 8 above)
Comment (capossele) [reply website/api/bonsai/bonsai-on-eth.md:26]: I decided to keep the more information we can on the docs and streamline the README

File: website/api_versioned_docs/version-0.19/bonsai/blockchain-zkvm-guide.md
Comment (capossele) [reply website/api_versioned_docs/version-0.19/bonsai/blockchain-zkvm-guide.md:1]: Nice catch!

File: xtask/Cargo.toml (hunk 2 above)
Comment (capossele) [reply xtask/Cargo.toml]: After removing the example from the Relay README, this change is not required anymore

Comment (capossele) [reply xtask/Cargo.toml]: Nevermind:
```
cargo run --bin cargo-risczero --no-default-features -- risczero install
  shell: /usr/bin/bash -e {0}
  env:
    CARGO_TERM_COLOR: always
    MSYS: winsymlinks:nativestrict
    GITHUB_TOKEN: ***
    Updating git repository `[https://github.com/paritytech/polkadot-sdk/`](https://github.com/paritytech/polkadot-sdk/%60)
    Updating crates.io index
    Updating git repository `[https://github.com/w3f/ring-vrf`](https://github.com/w3f/ring-vrf%60)
    Updating git repository `[https://github.com/w3f/fflonk`](https://github.com/w3f/fflonk%60)
    Updating git repository `[https://github.com/w3f/ring-proof`](https://github.com/w3f/ring-proof%60)
error: failed to select a version for `tempfile`.
    ... required by package `xtask v0.1.0 (/opt/actions-runner/_work/risc0/risc0/xtask)`
versions that meet the requirements `^3.5` are: 3.8.1, 3.8.0, 3.7.1, 3.7.0, 3.6.0, 3.5.0

all possible versions conflict with previously selected packages.

  previously selected package `tempfile v3.3.0`
    ... which satisfies dependency `tempfile = "~3.3"` of package `cargo-generate v0.18.0`
    ... which satisfies dependency `cargo-generate = "^0.18"` of package `cargo-risczero v0.19.0 (/opt/actions-runner/_work/risc0/risc0/risc0/cargo-risczero)`

failed to select a version for `tempfile` which could resolve this conflict
```

File: website/api/bonsai/bonsai-on-eth.md [hunk 13]
```diff
@@ -40,15 +40,15 @@ A typical flow works as follows:
...
+# use alloy_primitives::U256;
+# use alloy_sol_types::SolValue;
+# use anyhow::Context;
+# use bonsai_ethereum_relay::sdk::client::{CallbackRequest, Client};
+# use clap::Parser;
+# use ethers::{types::Address, utils::id};
+# use fibonacci_methods::FIBONACCI_ID;
+# use risc0_zkvm::sha::Digest;
 
 /// Exmaple code for sending a REST API request to the Bonsai relay service to
```
Comment (tzerrell) [inline website/api/bonsai/bonsai-on-eth.md:53]: Typo
```suggestion
/// Example code for sending a REST API request to the Bonsai relay service to
```

File: website/api_versioned_docs/version-0.19/bonsai/bonsai-on-eth.md [hunk 14]
```diff
@@ -2,18 +2,133 @@
...
+# use alloy_primitives::U256;
+# use alloy_sol_types::SolValue;
+# use anyhow::Context;
+# use bonsai_ethereum_relay::sdk::client::{CallbackRequest, Client};
+# use clap::Parser;
+# use ethers::{types::Address, utils::id};
+# use fibonacci_methods::FIBONACCI_ID;
+# use risc0_zkvm::sha::Digest;
+
+/// Exmaple code for sending a REST API request to the Bonsai relay service to
```
Comment (tzerrell) [inline website/api_versioned_docs/version-0.19/bonsai/bonsai-on-eth.md:53]: Typo
```suggestion
/// Example code for sending a REST API request to the Bonsai relay service to
```

**Review by tzerrell — APPROVED** [review]

File: .github/workflows/main.yml [hunk 15]
```diff
@@ -122,7 +122,10 @@ jobs:
       - uses: actions/checkout@v4
         with:
           lfs: true
+          # Recursive checkout required to build Solidity contracts and bindings.
+          submodules: recursive
       - uses: ./.github/actions/rustup
+      - uses: risc0/foundry-toolchain@2fe7e70b520f62368a0e3c464f997df07ede420f
```
Comment (flaub) [inline .github/workflows/main.yml:128]: Why is this needed for the `test` job? It should only be required for the `website-test` and `bonsai` jobs.

Comment (capossele) [reply .github/workflows/main.yml:128]: yeah I'm also kinda surprised, but this is what CI gets without:
```
cargo test -F $FEATURE -F profiler --workspace --exclude doc-test
error: failed to run custom build command for `bonsai-ethereum-contracts v0.5.0 (/opt/actions-runner/_work/risc0/risc0/bonsai/ethereum)`
note: To improve backtraces for build dependencies, set the CARGO_PROFILE_TEST_BUILD_OVERRIDE_DEBUG=true environment variable to enable debug information generation.

Caused by:
  process didn't exit successfully: `/opt/actions-runner/_work/risc0/risc0/target/debug/build/bonsai-ethereum-contracts-560e815577b67762/build-script-build` (exit status: 1)
  --- stderr
  Error: 
  Failed to resolve file: "/opt/actions-runner/_work/risc0/risc0/bonsai/ethereum/lib/forge-std/src/Test.sol": No such file or directory (os error 2).
   Check configured remappings..
      --> "/opt/actions-runner/_work/risc0/risc0/bonsai/ethereum/contracts/BonsaiCheats.sol"
          "forge-std/Test.sol"
  Error: `forge build` exited with failed status: exit status: 1
```
